### PR TITLE
daemon, client, cmd/snap: connect/disconnect now async

### PR DESCRIPTION
**Pull request by chipaca**

### CODE REVIEW

**Comment by chipaca** [comment]
Note this *will* break `client`. That'll come in a followup branch.

**Comment by niemeyer** [comment]
Tests pass, quite unfortunately.

File: client/interfaces.go [hunk 1]
```diff
@@ -78,21 +78,17 @@ func (client *Client) Interfaces() (interfaces Interfaces, err error) {
...
-		return err
-	}
-	return nil
+	return client.doAsync("POST", "/v2/interfaces", nil, bytes.NewReader(b))
 }
 
 // Connect establishes a connection between a plug and a slot.
 // The plug and the slot must have the same interface.
-func (client *Client) Connect(plugSnapName, plugName, slotSnapName, slotName string) error {
+func (client *Client) Connect(plugSnapName, plugName, slotSnapName, slotName string) (string, error) {
```
Comment (niemeyer) [inline client/interfaces.go]: Can we please name the results here and below, so docs tell easily what the first argument is?

File: cmd/snap/cmd_snap_op.go [hunk 2]
```diff
@@ -73,6 +73,9 @@ func wait(client *client.Client, id string) error {
 			return fmt.Errorf("change finished in status %q with no error message", chg.Status)
 		}
 
+		// note this very purposely is not a ticker; we want
+		// to sleep 100ms between calls, not call once every
+		// 100ms.
```
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Thanks! :)

**Comment by niemeyer** [comment]
LGTM

File: cmd/snap/cmd_connect.go [hunk 3]
```diff
@@ -68,5 +68,12 @@ func (x *cmdConnect) Execute(args []string) error {
...
 	}
-	return Client().Connect(x.Positionals.Offer.Snap, x.Positionals.Offer.Name, x.Positionals.Use.Snap, x.Positionals.Use.Name)
+
+	cli := Client()
+	id, err := cli.Connect(x.Positionals.Offer.Snap, x.Positionals.Offer.Name, x.Positionals.Use.Snap, x.Positionals.Use.Name)
+	if err != nil {
+		return err
+	}
+
+	return wait(cli, id)
```
Comment (zyga) [inline cmd/snap/cmd_connect.go]: Wow, this is pretty nice indeed!

File: cmd/snap/cmd_connect_test.go [hunk 4]
```diff
@@ -62,24 +62,31 @@ Help Options:
 
 func (s *SnapSuite) TestConnectExplicitEverything(c *C) {
 	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
```
Comment (zyga) [inline cmd/snap/cmd_connect_test.go]: Ah, quite interesting. So we POST, get an change ID (``zzz``) and then we poll that change by GETting a 2nd URL, right?

File: daemon/api_test.go [hunk 5]
```diff
@@ -1468,16 +1468,24 @@ func (s *apiSuite) TestConnectPlugSuccess(c *check.C) {
...
-	})
+	id := body["change"].(string)
+
+	st := d.overlord.State()
+	st.Lock()
+	chg := st.Change(id)
+	st.Unlock()
+	c.Assert(chg, check.NotNil)
+
+	<-chg.Ready()
```
Comment (zyga) [inline daemon/api_test.go]: Nice, I wasn't aware of Change.Ready!

**Comment by zyga** [comment]
+100 :-) Thanks. Reading the diff was very informative.

File: cmd/snap/cmd_connect_test.go (hunk 4 above)
Comment (chipaca) [inline cmd/snap/cmd_connect_test.go]: for now, yes. At some point we'll get the websocket working, and then it'll be: POST, connect to websocket, wait on it

**Pull request merged**